src/Pate/Memory/MemTrace.hs

```diff
@@ -500,22 +506,43 @@ instance TestEquality (SymExpr sym) => Eq (MemOp sym ptrW) where
      , valR == valR'
      , Just Refl <- testEquality valO valO'
     = cond == cond' && dir == dir' && end == end'
-  MergeOps p opsT opsF == MergeOps p' opsT' opsF'
-    | Just Refl <- testEquality p p'
-    = opsT == opsT' && opsF == opsF'
+
   _ == _ = False
 
+instance OrdF (SymExpr sym) => Ord (MemOp sym ptrW) where
+  compare (MemOp (LLVMPointer reg1 off1) dir1 cond1 sz1 (LLVMPointer vr1 vo1) end1)
+          (MemOp (LLVMPointer reg2 off2) dir2 cond2 sz2 (LLVMPointer vr2 vo2) end2) =
+    case compareF sz1 sz2 of
+      LTF -> LT
+      GTF -> GT
+      EQF ->
+        (compare reg1 reg2) <>
+        (toOrdering $ compareF off1 off2) <>
+        compare dir1 dir2 <>
+        (compare cond1 cond2) <>
+        (compare vr1 vr2) <>
+        (toOrdering $ compareF vo1 vo2) <>
+        compare end1 end2
+
+data MemEvent sym ptrW where
+  MemOpEvent :: MemOp sym ptrW -> MemEvent sym ptrW
+  SyscallEvent :: forall sym ptrW w.
+    -- TODO, something more realistic
+    SymExpr sym (BaseBVType w) ->
+    MemEvent sym ptrW
+
 data MemTraceImpl sym ptrW = MemTraceImpl
   { memSeq :: MemTraceSeq sym ptrW
-  -- ^ the sequence of memory operations in execution order
+  -- ^ The sequence of memory operations in reverse execution order;
+  --   later events appear closer to the front of the sequence.
   , memState :: MemTraceState sym ptrW
   -- ^ the logical contents of memory
   }
 
 data MemTraceState sym ptrW = MemTraceState
   { memArr :: MemTraceArr sym ptrW }
 
-type MemTraceSeq sym ptrW = Seq (MemOp sym ptrW)
+type MemTraceSeq sym ptrW = SymSequence sym (MemEvent sym ptrW)
```

Comment:
Ah yes - I remember now. I had an experimental branch where I made this change too (but it was in a copy of the memory model I was experimenting with). I think this is entirely reasonable.

Reply:
For reference, that PR is here https://github.com/GaloisInc/pate/pull/174/files#diff-f9510819c7bea75a6da4de8da706d3d6767f09b84e4539040629aef5a4b369dc. I tried to split up this monolithic MemTrace.hs module into more logically separate components

Reply:
No... the two dimensional array memory model idea did not pan out for performance reasons.

Maybe I'll resurrect that module split later